Implement a `help()` method for the configuration system

eerovaher opened the pull request:
`ConfigItem` instances have a `description` attribute, the contents of which are written to the configuration file as comments. However, there is currently no convenient way of accessing the `description` strings through the configuration system at runtime (see the awkward code examples in the documentation that this pull request replaces). The new `ConfigNamespace.help()` method provides a user-friendly way of seeing the name, current value, type and description of one or more configuration items.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

taldcroft commented:
I haven't looked at the code in detail but this is an awesome idea. 💯

mhvk commented:
I like the idea too, but wonder if a simpler solution to the problem is not to define a `__repr__` which essentially includes your help text for all names.

mhvk commented:
Actually, just because I wondered what would happen, `help(conf)` is quite good...

```
from astropy.utils.iers import conf

help(conf)

Help on Conf in module astropy.utils.iers.iers object:

class Conf(astropy.config.configuration.ConfigNamespace)
 |  Configuration parameters for `astropy.utils.iers`.
 |  
 |  Method resolution order:
 |      Conf
 |      astropy.config.configuration.ConfigNamespace
 |      builtins.object
 |  
 |  Data descriptors defined here:
 |  
 |  auto_download
 |      Enable auto-downloading of the latest IERS data.  If set to False then the local IERS-B file will be used by default (even if the full IERS file with predictions was already downloaded and cached). This parameter also controls whether internet resources will be queried to update the leap second table if the installed version is out of date. Default is True.
 |  
 |  auto_max_age
 |      Maximum age (days) of predictive data before auto-downloading. See "Auto refresh behavior" in astropy.utils.iers documentation for details. Default is 30.
 |  
 |  iers_auto_url
 |      URL for auto-downloading IERS file data.
 |  
 |  iers_auto_url_mirror
 |      Mirror URL for auto-downloading IERS file data.
 |  
 |  iers_degraded_accuracy
 |      IERS behavior if the range of available IERS data does not cover the times when converting time scales, potentially leading to degraded accuracy.
 |  
 |  iers_leap_second_auto_url
 |      URL for auto-downloading leap seconds.
 |  
 |  ietf_leap_second_auto_url
 |      Alternate URL for auto-downloading leap seconds.
 |  
 |  remote_timeout
 |      Remote timeout downloading IERS file data (seconds).
 |  
 |  system_leap_second_file
 |      System file with leap seconds.
 |  
```

eerovaher commented:
> Actually, just because I (EDIT: thought wouldn't it be nice if one could just do "help(something)" and) wondered what would happen, `help(conf)` is quite good...

Python's builtin `help()` function prints quite a lot for the `ConfigNamespace`, but if help is wanted for a single item then this could be overwhelming. The `ConfigNamespace.help()` method I am proposing would allow for printing an overview of a single configuration item, and it shows more information per item than the builtin `help()` does. In particular, if a configuration item accepts only a finite set of values then `ConfigNamespace.help()` lists all of them.

> I like the idea too, but wonder if a simpler solution to the problem is not to define a `__repr__`

`ConfigItem.__str__()` already produces an informative string:

https://github.com/astropy/astropy/blob/f2c67ef6f0fa4c81406698b0b711541d940d2d71/astropy/config/configuration.py#L394-L404

However, thanks to `ConfigItem.__get__()` accessing the configuration items normally (as attributes of the `ConfigNamespace` instance) retrieves their value, not the `ConfigItem` instance, so the `ConfigItem.__str__()` method is quite difficult to access.

I have updated the pull request so that if help is printed for multiple configuration items then the entries are separated with empty lines, and there is now an example in the docstring.

mhvk commented:
I guess we should separate the two concerns, but right now, while the configitems may have a nice `__str__` method, the configuration entity does not: `print(conf)` and `repr(conf)` are both essentially useless! I think this is partially a question of what users are most likely to discover; I think that starts with just giving something useful for the `repr` -- most `conf`s do not have that many entries, so generally it will not be overwhelming.  

Note that in `ipython`, `conf?` is also not very helpful (`conf??` is better but less good than `help(conf)`.

All that said, I definitely see your point that just from tab-completion, it is nearly impossible to get useful information; e.g., `conf.auto_download?` gives information on bools... My main worry is just that it is unlikely any user would guess there would be a `.help()` method (though presumably they might find it through tab completion); for the somewhat related `Table.read.help`, the main trick was to put the existence of the `.help` method high up in the docstring (though I still worry I only know about it because I saw the PR!).

eerovaher commented:
I've implemented `ConfigNamespace.__str__()`. The string representation of a `ConfigNamespace` contains the string representations of the configuration items within it, and is used in `ConfigNamespace.help()`. Although calling `print(conf)` now results in the same message as calling `conf.help()` without arguments, having the `help()` method is still useful because it also allows printing the string representation of a single configuration item.

I have converted this pull request to a draft so that we could discuss if the `ConfigItem.__str__()` should be changed or if we are happy with the way the string representation of a configuration item is currently formatted.

eerovaher commented:
@mhvk, does the proposed implementation address your remarks?

eerovaher commented:
@mhvk, could you have a look at this? I converted this pull request to a draft in case we want to also change `ConfigItem.__str__()`, but if we don't then this should be good to merge as is.

mhvk commented:
I'm so sorry that this keeps slipping by. The problem is mostly that really all I remember is a feeling of disquiet at this, but I know that I actually have to look at things again, for which I don't really have time right now. It may be more logical is some of the people who wrote the configuration system had a look, so let's ping @eteq, @saimn and @WilliamJamieson .

eerovaher commented:
I'm trying to add a change log entry, but even though I can successfully update the correct branch in my fork the pull request does not get updated. I'll have to try closing and reopening the pull request.

pllim commented:
GitHub having issues right now. You might have to wait it out.

eerovaher commented:
The pull request now matches my branch, so this is ready for review.

A short summary of this pull request is that it implements `ConfigNamespace.__str__()` and `ConfigNamespace.help()`, both of which make it simple to learn about configuration items from within Python. However, both implemented methods make use of `ConfigItem.__str__()`, which I haven't edited, but perhaps it should be updated so that the output of `ConfigNamespace.help()` would be more convenient.

pllim commented:
I am interested to review but unfortunately can't get to this in a week or a few...

eerovaher commented:
I rebased this once more as requested by https://groups.google.com/g/astropy-dev/c/9DO4YkoU44g

pllim commented:
codecov upload failed, so we can ignore the coverage complain.

pllim commented on docs/config/index.rst:
```diff
@@ -220,22 +220,27 @@ To see what configuration parameters are defined for a given ``conf``::
-    >>> conf.auto_max_age
-    30.0
 
-You can also iterate through ``conf`` in a dictionary-like fashion::
```
Maybe we still want to advertise that one can iterate through like `dict`. I vaguely remember having to do such a thing a long time ago for other reason, so I think this is still good to document even if you don't need it for description anymore.

pllim commented:
Despite using a `.help()` method, unlike typical Python `str` or `repr`, I still think this is an improvement based on the doc updates. Just have a small comment on the example that was removed.

pllim commented:
I think this also works for downstream packages that uses the config system (e.g., `synphot`), so that is nice. 👏

pllim reviewed (approved):
Sorry this dragged on so long. Thanks!
